### ISSUE TITLE
Support CSS Modules

### ISSUE THREAD

**Issue by kitsonk** [denoland/deno#11961]
Chromium is shipping CSS Modules unflagged in v93: https://chromestatus.com/feature/5948572598009856

We should consider how this applies to Deno.

The relevant commit to the HTML Living Spec: https://github.com/whatwg/html/commit/3d45584d286e9455cc24ebae1f3aca3db120dc9d

The biggest challenge is that a constructed [`CSSStyleSheet`](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet) object is returned on the import which then is inserted into the DOM.  Deno doesn't have a DOM or the `CSSStyleSheet` object, nor is it clear what expectations would be for users and varies greatly from what a front-end bundler would consider a "CSS module".

Personal opinion is that everything is very vague that it really is unimplementable.  (Plus we are also blocked on having import assertions, which are still not supported by tsc)

**Comment by jtoppine** [comment]
CSSStyleSheet and friends sound like out of scope for deno perhaps better suited to a third-party module, but having low level access to CSS modules in deno would be useful. I can imagine a use case to only need:

1) ability to programmatically inspect and read (just raw text data) CSS imports of any given module and it's dependencies
2) CSS and JSON imports being bundled/compiled alongside js. 

When the same JS modules are used both for server side rendering as well as client side browser code, this would enable server applications to somehow collect the CSS modules used and present them to the client as a traditional, pre-baked CSS file when SSR is being used. 

Frameworks or server apps would be free to implement this kind of functionality however they best see fit, but I can imagine it being desireable for many to have CSS dependencies managed alongside JS.

**Comment by Macil** [comment]
For anyone else who is confused like I was: the CSS Modules feature talked about here (sometimes disambiguated as "CSS Module Scripts") **is unrelated to** the popular [CSS Modules feature](https://github.com/css-modules/css-modules) (`import styles from "styles.module.css";`) supported by [webpack/css-loader](https://github.com/webpack-contrib/css-loader), [Vite](https://vitejs.dev/guide/features.html#css-modules), [Parcel](https://parceljs.org/languages/css/), [Next.js](https://nextjs.org/docs/basic-features/built-in-css-support), [Create React App](https://create-react-app.dev/docs/adding-a-css-modules-stylesheet/), the Deno framework [Aleph.js](https://alephjs.org/), etc, and **it's also unrelated to** all of those tools' ability to import non-module CSS files to load them as regular global stylesheets (`import "styles.css";`).

The CSS Module Script standard isn't used by any popular framework to my knowledge, implementing it in Deno would not help projects that want to use CSS Modules, and it doesn't seem to be built for server-side rendering.

From https://github.com/WICG/webcomponents/issues/759#issuecomment-520566796, with some clarifying details and emphasis added:

>- ICSS modules [referring to the pre-existing .module.css CSS Modules] work great with server-side JS, allowing developers to return plain HTML + CSS with no client-side JS required for rendering. I can't see how a CSSStyleSheet [used in CSS Module Scripts] could be used on the server side at all.
>- ICSS modules are popular, and developers have been asking (perhaps naively) for native browser support for "CSS Modules" for years, in the same naive way that they wish for JSX built-in to the browser. **This proposal [CSS Module Scripts] is not at all what they meant.**

Not that I'm against this being implemented in Deno just because it's not CSS Modules -- it wouldn't get in the way of them -- but I just want to emphasize that it's unclear how it would be used and who would benefit if it was implemented in Deno specifically.

Implementing this would require implementing `CSSStyleSheet` from the CSS Object Model, which is related to the DOM, and it seems like Deno has mostly treated the DOM as out of scope. I think this combined with the likely lack of a use-case would be a good reason to not implement it in Deno. (If possible, it might be neat to make it so the DOM and CSSOM could be implemented in third-party modules in order to simulate a browser DOM in Deno maybe for testing purposes, and then provide the proper import hooks so CSS Module Scripts could be fully implemented through a third-party module.)

**Comment by ayame113** [comment]
I think #13898 explains how `CSSStyleSheet` is useful in Deno.
If `CSSStyleSheet` is implemented in Deno, it would be convenient to have CSS import. +1

**Comment by timreichen** [comment]
+1

**Comment by jespertheend** [comment]
I’m using Deno as a test runner for code that normally runs in the browser. So Deno simply not throwing when it encounters a css import assertion would already be enough for that specific use case.

But I realize throwing is probably preferable rather than to have it fail silently with undefined behavior.

**Comment by ayame113** [comment]
On a separate note, I'm having trouble type checking code for browsers that have CSS imports.

Currently, I create `.d.ts` and perform type checking in the following way.

```ts
// mod.ts

// @deno-types="./stylesheet.d.ts"
import css from "./style.css";
console.log(css);
```

```ts
// stylesheet.d.ts

/// <reference lib="dom" />
declare const stylesheet: CSSStyleSheet;
export default stylesheet;
```

With this method, type checking works as expected in the editor,

![image](https://user-images.githubusercontent.com/40050810/214481974-d90a378c-2a61-4ea5-94b2-b843bfb466b5.png)

but an error occurs in the `deno check` command.

```shell
> deno check ./mod.ts
error: The import assertion type of "css" is unsupported.
    at file:///C:/Users/
```

It may take some time to implement the `CSSStyleSheet` API in the future, but it would be nice if there was a way to do type checking with the `deno check` command before that.

**Comment by bartlomieju** [comment]
@dsherret it looks like a `deno_graph` thing. I think it's reasonable to allow `css` assertion there. WDYT?

**Comment by aapoalas** [comment]
Fun related spec issue: https://gist.github.com/peetklecha/a55532165dbd4905aa91bbe59e8b1001

TL;DR: Import assertions will be going back to stage 2 and the syntax will likely change.

**Comment by jespertheend** [comment]
As a workaround, I ended up adding an entry to Deno's import map. All my css imports were already contained in a single module. So now the import map maps this module to a dummy module that exports empty objects instead:

```js
// styles.js

// @ts-ignore
import mySheet from "./mySheet.css" assert {type: "css"};
export {mySheet};
```

```js
// denoStyles.js

export const mySheet = {};
```

**importmap.json**
```
{
	"imports": {
		"./path/to/styles.js": "./path/to/denoStyles.js"
	}
}
```

The browser doesn't have this entry in its import map, so it loads `browserStyles.js` like before, whereas Deno maps it to `denoStyles.js` which doesn't cause any errors.

**Comment by Mqxx** [comment]
Any update on this?

**Comment by mattvr** [comment]
Looks like import assertions are now Stage 4: https://github.com/tc39/proposal-import-attributes. This would be really useful for web development w/ Deno.

**Comment by petamoriken** [comment]
I assume that the CSS module is not implemented due to low priority. If I can make time for it, I would consider implementing it.

**Comment by bartlomieju** [comment]
@petamoriken we discussed this briefly among the team and we are not yet sure if we want to add this API. It appears that community "CSS modules" (https://github.com/css-modules/css-modules) is much more popular than the CSS Modules that Chrome has shipped, that are mainly used for Web Components. Please hold off before implementing it.

**Comment by dkolba** [comment]
@bartlomieju using css via import attributes with constructable stylesheets and `@scope` within the stylesheet is 95% feature equivalent to community CSS Modules, plus it works out of the box without bundlers in the browser (well, Chrome that is)...

**Comment by landsman** [comment]
CSS modules are awesome and still simple - just fixing annoying issue with naming. Please consider them.

**Comment by petamoriken** [comment]
> I don't want to have native support in Deno. Just do not block to use them via Vite plugin like this https://github.com/privatenumber/vite-css-modules

I think non native CSS modules from ecosystems like Vite are already available without any specific Deno support, is that wrong?

**Comment by landsman** [comment]
> I think non native CSS modules from ecosystems like Vite are already available without any specific Deno support, is that wrong?

@petamoriken Do you have some example repo? I wasn't able to make it work (vite+react+css modules).

**Comment by zerodegress** [comment]
> > I think non native CSS modules from ecosystems like Vite are already available without any specific Deno support, is that wrong?
> 
> [@petamoriken](https://github.com/petamoriken) Do you have some example repo? I wasn't able to make it work (vite+react+css modules).

just add this abobe importing css module:

```
// @ts-ignore css modules
```

it works, but i thought deno could provide a better solution for it.

**Comment by Mqxx** [comment]
Any update on this? 🙂

**Comment by bavoco** [comment]
Looks like Firefox is planning on shipping (web standard) css modules: 
https://bugzilla.mozilla.org/show_bug.cgi?id=1720570#:~:text=the%20development%20work%20should%20finish%20this%20year

**Comment by shine-on-me** [comment]
4 years later... time for a little workaround if anyone's interested. This is for a use case were you'd want to render some web component on the client and for SSR purposes on Deno as well without rewriting the whole component from scratch.

Basically I got it working with a nasty workaround by using Deno's text imports. So you'd have to copy your CSS file, make it a TXT file and add a little bit of extra logic on the browser side. So the tradeoff would be one extra .txt file per .css file and a little extra logic on the browser side...Not great but not terrible considering the lack of alternatives.

This assumes a buildless setup on the browser side using import maps with Lit (hence the JSDOC) and Deno as package manager on the server side. I'd say the general logic applies regardless of setup, JS framework etc.

```js
// example.js

/**
 * @file example.js
 * @description Buildless web component for the browsers, but we ALSO want to SSR this on our Deno server!
 * 
 * Add importmap to your html <head> for using lit buildless like this:
 * (esm.sh is used as an example but you can bundle whatever you want)
     <script type="importmap">
         {
             "imports": {
                 "lit": "https://esm.sh/lit@3.3.1"
             }
         }
     </script>
 */

// need unsafeCSS so Deno stays happy later on
import { LitElement, unsafeCSS, html, isServer } from "lit"

// Nice and easy for modern browsers, Deno 2.5.3 actively blocks this and no support for CSSStylesheets :(
// import style from "./example.css" with { type: "css" }

// Temporary workaround that works for modern browsers AND Deno 2.5.3
const { default: importedCSS } = isServer
	? // For Deno just rawdog text
		await import("./example.txt", { with: { type: "text" } })
	: // @ts-ignore BS, importing CSS is supported in browsers now (Firefox needs temporary polyfill but w/e)
		await import("./example.css", { with: { type: "css" } })

class MyExample extends LitElement {
	/**
	 * @override
	 * Add styles regardless of environment
	 * On the server importedCSS will be a string, in the browser a CSSStylesheet
	 * Lit has isServer, but you can apply your own environment logic here if needed
	 */
	static styles = isServer ? unsafeCSS(importedCSS) : importedCSS

	constructor() {
		super()
	}

	/**
	 * @override
	 * Fully CSS styled (!) Lit web component where we render whatever
	 */
	render() {
		return html`<p>My styled example component</p>`
	}
}

customElements.define("my-example", MyExample)

```

And then in Deno you can do something like this, just be sure to run with `--unstable-raw-imports` or add `"unstable": ["raw-imports"]` to your `deno.jsonc`.
```ts
// ssr.ts

import { render, html } from "@lit-labs/ssr"
import { collectResult } from "@lit-labs/ssr/lib/render-result.js"

// your browser code, in our case a LitElement so we can use Lit's SSR library
import "example.js"

const template = html`<my-example></my-example>`
const renderRes = render(template)
const ssr = await collectResult(renderRes)

// Here we get the exact same output as on the browser, including our CSS ;D
console.log(ssr)
```

**Comment by CertainLach** [comment]
Native CSS modules were shipped in firefox: https://www.firefox.com/en-US/firefox/147.0/releasenotes/

**Comment by Alastyr** [comment]
Hi everyone. Any news about this issue?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#loading_non-javascript_resources

**Comment by csvn** [comment]
Wow, I didn't realize the support was this far along, with [Webkit, Deno & NodeJS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import/with#browser_compatibility) being the ones left to implement this feature.

**Comment by bartlomieju** [comment]
Done in #35093 and released in Deno v2.9: https://deno.com/blog/v2.9#css-module-imports
